src/main/java/com/cloudbees/plugins/credentials/impl/CertificateCredentialsImpl.java

```diff
@@ -129,17 +129,22 @@ public class CertificateCredentialsImpl extends BaseStandardCredentials implemen
     public CertificateCredentialsImpl(@CheckForNull CredentialsScope scope,
                                       @CheckForNull String id, @CheckForNull String description,
                                       @CheckForNull String password,
-                                      @NonNull KeyStoreSource keyStoreSource) {
+                                      @NonNull KeyStoreSource keyStoreSource) throws Descriptor.FormException {
```

Comment:
This signature change in a public method might break some plugins --> https://github.com/search?q=org%3Ajenkinsci+%22new+CertificateCredentialsImpl%22&type=code

Reply:
Yes, the change is binary-compatible but source-incompatible. The [Descriptor#newInstanceImpl](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/Descriptor.java#L623) throws a LinkageError if an IllegalArgumentException is thrown from the constructor. Using exceptions like Descriptor.FormException ensures that errors are rendered with a helpful error message instead of the angry Jenkins page.

Please let me know if there is a better approach.

Reply:
Removed the incompatible changes and handled the LinkageError in CredentialsSelectHelper#doAddCredentials